### PR TITLE
feat: a false sentence break is repaired by choosing between readings

### LINKED ISSUE #260: Score three readings with Qwen instead of a threshold, to repair 82% of false sentence breaks instead of 26%
**Depends on:** nothing. **Blocks:** #256 (easier after this), and with #261 it removes ModernBERT from the app.

A pause mid-sentence makes the recogniser write a period and capitalise the next word. The stage that repairs this fixes 26% of them, and it cannot be tuned higher, because its threshold is set by the single lowest-scoring real sentence ending out of 180. Replacing the score with a choice between readings fixes 82% and calibrates nothing. Measured in Swift with the 4-bit model that would ship: 50 ms median per boundary, zero wrong joins on 180 real endings.

Today `SentenceProbe.at()` masks the boundary and computes `log P(".") − log P(" next word")`, and `Config.swift:1674-1675` compares it to `joinBelow: -4` and `offerBelow: -2`. Instead, build three readings of the same text, score each as a whole sequence per token, and take the winner.

<details>
<summary>The calculation, exactly</summary>

For a `word. Capital` boundary, take up to 12 words either side. Build three
readings and score each with a causal model:

    ". Term A is longer"     period, next word capitalised
    "  term A is longer"     joined, next word lowercased
    ", term A is longer"     comma, next word lowercased

Score = sum of `log P(token | everything before it)` over the continuation,
**divided by the token count**. Highest wins. If `joined` wins, remove the
period and lowercase the next word — the repair itself is unchanged.

Three things are load-bearing:

**Per token, not summed.** On summed log-probability the joined reading wins by
being shortest: 97% repairs and 33 wrong joins.

**The comma is an option, not a term in a subtraction.** Scoring
`max(. , ; :) − log P(next)` gives 64% against the shipped shape's 81%, because
a comma is plausible everywhere and lifts the cuts with the real endings. As a
*reading* it is essential: 26% of real endings pick the comma, which is why none
of them picks `join`. Those are the boundaries the user labelled by hand as
"really a comma".

**`;` and `:` never change a decision in English.** Three readings, not five.
Keep the mark set in config — it is language-dependent, and French uses `:`
where English does not.

</details>

<details>
<summary>Measured — 26% to 82%, zero wrong joins</summary>

Bench: 172 real sentence endings and 140 spurious cuts, from the user's own
dictation. `sentence-join/harness/en_real_nf.json`, `en_cuts_nf.json`,
`en_real_labels.json`.

    shipped ModernBERT at -4          26% repaired    0 wrong joins /172
    ModernBERT at its best floor      81%             0    <- floor fitted to this data
    Qwen3-0.6B-Base fp32              79%             0
    Qwen3-0.6B-Base-4bit  [ship]      82%             0

The 81% only exists because the user relabelled the bench by hand on
2026-09-02; before that it was 67%. It is a number fitted to 172 cases and set
by the worst one, so it does not transfer. The argmax fits nothing.

4-bit versus fp32: 47 of 324 argmax decisions differ, but 39 are `.` versus `,`
and both mean leave it alone. Only 8 change the answer, 6 of them gains.

</details>

<details>
<summary>Latency, measured in Swift — and why not to build a KV cache</summary>

M2 Pro, Ollama holding 9.6 GB on the GPU, load average 5.5–10.

    three readings in one padded batch    median 50.1 ms   p95 52-60 ms
    naive, prefix re-run per reading      median 70.0 ms
    prefix KV cache                       median 88.1 ms   <- slower
    one forward alone                     median 23.5 ms

**Batch the readings. Do not build a KV cache.** One forward costs 23.5 ms
regardless of length at 0.6B — the pass is bound by reading the weights, not by
the sequence — so caching the shared prefix adds a fourth pass and saves
nothing. The cache path is also *wrong*: 2 of 972 sequences retokenise their
prefix (the case is `names).`), the shared cache then holds the wrong tokens,
and it produces one wrong join. The same trap sits under any shared-prefix
optimisation.

Load 1.3 s warm, first decision 105–113 ms including Metal kernel compile,
weights 320 MB, peak 526 MB. As a third model it needs
`MLX.GPU.set(cacheLimit:)` — MLX's buffer pool grows to 1–2 GB across varied
shapes.

Full run: `~/Documents/parrotflow-scratch/qwen-timing/RESULT.md`.

</details>

<details>
<summary>The checkpoint — not the obvious one</summary>

Use **`mlx-community/Qwen3-0.6B-Base-4bit`**.

NOT `mlx-community/Qwen3-0.6B-4bit-DWQ`. Despite the name it is a quant of the
**instruct** model — `eos_token_id: 151645` — and scores 76% against Base's 82%,
with 18 join-flips of which 11 are losses.

`Package.swift:15` already depends on `mlx-swift-lm` 3.31.4 and
`Sources/ParrotFlow/WordVectors.swift:30` already downloads and loads a 4-bit
MLX Qwen. Read `WordVectors.swift` before writing the loader — it solved the
download path, and its comment at line 32 records that
`model.safetensors.index.json` is absent from these repositories. It uses the
*embedding* checkpoint for hidden states; this needs the **base** model's
logits, which is a different code path (`MLXLLM` causal LM).

</details>

<details>
<summary>What is still unmeasured</summary>

**Whether stripping fillers first still helps.** `Transcriber.swift:487` runs
this stage *before* the `fillers` and `repetitions` pipeline steps, so the
scorer sees `um` and `the the`. Stripping them before scoring was worth 14
points to the old *threshold* — it moved the safe line from -0.23 to +0.84. With
an argmax there is no line to move, so the value is unknown and may be zero. One
cheap re-run against `harness/qmarks.json` settles it. Do that before deciding
whether the strip belongs in this PR.

**The cheap-screener cascade.** Running the argmax only when a masked score sits
in (-4, 4) covers 44% of boundaries with no loss. But ModernBERT is leaving, and
mmBERT-small scores 47% on the boundary probe, so whether it can screen is
untested. Not needed at 50 ms; worth knowing if the cost ever matters.

**The collocation case.** Joining can create a word that was not there:
`I would say no. Comment on the PR` becomes `no comment`, which inverts the
meaning. Constructed 16 such cases; only 2 are genuinely ambiguous, and the
other 14 read as one sentence anyway so joining them is correct. 0 of 180 real
endings are of this shape. Small and cheap, but it is the failure mode nobody
would notice, because a wrong join reads *more* fluently than what was said.

</details>

### PR DESCRIPTION
**Pull request by znat**
A pause mid-sentence makes the transcriber write a period and capitalise the next word. The stage that repairs this scored the boundary with ModernBERT and compared the score to two thresholds. It repaired 26% of the cuts, and the threshold could not be raised: it was set by the single lowest-scoring real sentence ending out of 180. The boundary is a choice now. Every `word. Capital` boundary is read once per mark and once with no mark at all, each reading is scored per token by `mlx-community/Qwen3-0.6B-Base-4bit` (320 MB, new download), and the highest wins. On the user's own dictation that repairs 81% of 140 cuts and joins none of 172 real endings, at 50 ms median per boundary. The new setting is `transcription.sentences.marks`, default `[".", ","]`; `sentences: false` still turns the stage off.

Start the review at `SentenceReadings.swift`. Check three things there: the readings are scored **per token**, not summed; the comma is a reading and not a term in a subtraction; and the guard in `score` that handles a prefix which retokenises inside the full sequence. Then check the fail-open path in `SentenceJoin.apply` — a dictation that arrives before the weights are in memory keeps its boundaries and starts the load, it never waits.

Closes #260.

<details>
<summary>Measured through the binary — 81% of 140 cuts repaired, 0 wrong joins of 172</summary>

Bench: `~/Documents/parrotflow-scratch/sentence-join/harness/en_real_nf.json` (194 real endings) and `en_cuts_nf.json` (130 cuts), with `en_real_labels.json` applied as `qwen-timing/eval_argmax.py` applies it — `join` moves to the cut set, `drop` and `tie` are removed. Run through `--sentence-probe --bench`, which is this PR's own code.

| | repaired | wrong joins |
|---|---|---|
| shipped ModernBERT at `join_below: -4` | 26% | 0 of 172 |
| **this PR, Qwen3-0.6B-Base-4bit, three readings batched** | **114 of 140 (81%)** | **0 of 172** |

**81%, not the issue's 82%.** The issue's headline is the naive arm, which runs one forward pass per reading. This PR batches the three readings into one padded pass, which is 1.4x faster and repairs one cut fewer. `qwen-timing/RESULT.md` §4 records the same: 114 batched against 115 naive, 0 wrong joins either way.

Row-by-row against the reference `swift_qmarks_base4bit.json` `batched` field: 323 of 324 boundaries agree to the last digit (max |delta| 0.00000). The one that differs is `real i=5`, and it differs because of the retokenisation guard — see the block below.

</details>

<details>
<summary>Stripping fillers first buys nothing now — the pipeline order stays</summary>

The issue asked whether the stage should run after the `fillers` step instead of before it. Stripping fillers was worth 14 points to the old threshold, because it moved the safe line. An argmax has no line to move, and the measurement says the value is zero.

Both file pairs scored through the same binary:

| | repaired | wrong joins |
|---|---|---|
| `*_nf.json`, fillers stripped | 114 of 140 (81%) | 0 of 172 |
| `en_real.json` / `en_cuts.json`, fillers kept | 114 of 140 (81%) | 0 of 172 |

5 of 324 decisions differ, and all five are `.` against `,`. Both mean leave the boundary alone, so no text changes. No join decision differs.

Worth knowing why the effect is so small: only 22 of the 324 boundaries have any filler in their ±12-word window at all. The two file pairs are identical text on the other 302.

**Recommendation: leave the order as it is.** Do not move this stage after `fillers` for the sake of the boundary decision.

</details>

<details>
<summary>The retokenisation guard fired on 1 of 324 boundaries, and cost nothing</summary>

`tokenize(prefix + continuation).prefix(p)` is not always `tokenize(prefix)`. On 2 of 972 bench sequences the boundary character merges into the last prefix token. The case is a left side ending `names).`. Slicing the continuation at the prefix's length then scores the wrong tokens.

`SentenceReadings.score` compares each full sequence against the prefix ids and starts the scored span at the first token that actually differs. It fired once, on `real i=5`:

    stored (unguarded)   .  -6.1517   ,  -6.2821   join  -6.6884
    guarded              .  -5.6512   ,  -5.6379   join  -6.6884

The argmax flips from `.` to `,`. Both mean leave the boundary alone, so 0 wrong joins of 172 holds. `join` is unchanged and still loses.

It is belt-and-braces in the app: `SentenceJoin.boundaries(in:)` already skips a period whose preceding character is not a letter or a digit, so `names).` would never reach the scorer. The guard is there because the trap sits under any prefix-sharing optimisation, and the next person to try one should find it handled.

</details>

<details>
<summary>Latency and memory — 50 ms median, and why the buffer pool is capped at 256 MB</summary>

M2 Pro, 32 GB, macOS 27.0. Ollama resident, other agents running on the same machine, so the load average moved between runs and is given with each number. `--sentence-probe --bench`, release build, the first decision dropped because it pays the Metal kernel compile.

| load average | models resident | boundaries | median | p95 |
|---:|---|---:|---:|---:|
| 14.7 | readings only | 194 real | 50.9 ms | 88.1 ms |
| 14.7 | readings only | 130 cuts | 56.0 ms | 87.6 ms |
| 10.9 | readings + word vectors | 130 cuts | 48.8 ms | 85.7 ms |
| 5.6 | readings + word vectors | 130 cuts | 36.1 ms | 50.9 ms |

The reference run measured 50.1 ms median at load 5.5–10, so this reproduces. The p95 is the tail a contended GPU gives.

Memory, both MLX models in one process, after 130 boundaries:

| buffer-pool limit | MLX active | MLX cache | MLX peak | process peak footprint | median |
|---|---:|---:|---:|---:|---:|
| 256 MB (this PR) | 639 MB | 256 MB | 837 MB | 1.25 GB | 36.1 ms |
| 2 GB (effectively none) | 639 MB | 1038 MB | 838 MB | 1.87 GB | 36.1 ms |

256 MB holds 620 MB less and costs nothing measurable. `WordVectors` set no limit before this PR, so the cap now lives in `MLXModelCache.limitBufferPool()` and is applied by whichever MLX model loads first.

Weights on disk 320 MB. Load 1.3 s warm; nothing waits for it.

</details>

<details>
<summary>Deliberate removals, and one that goes beyond the issue</summary>

**The offer tier.** `Tier.offer` is gone. An argmax has no threshold to hedge with, so there is no middle to report. `Tier` is now `join` and `leave`.

**`join_below` and `offer_below`.** Both keys are still parsed, and `--check-config` reports each as a notice:

    · sentences: `join_below:` was a threshold on a score. The boundary is settled
      by which reading the model scores highest now, so there is nothing left to set

The config still loads and the stage still runs. `sentences: false` and `sentences: {…}` are both still valid. Two shapes of `marks:` are refused. An empty list, because with no mark the join is the only reading and every period in the transcript would go. And anything that is not a single punctuation character: a mark is written into the sentence as punctuation, and the word `join` is the name of the reading that removes the period, so `marks: ["join"]` would have removed one.

**`sentences.offered`** is removed from the pipeline scope. Nothing in `docs/`, `tests/` or `scripts/` read it. `sentences.joined` stays.

**Beyond what the issue asked for: `SentenceProbe.read()`, its `Reading` struct, `Failure.empty` and the `period` field are deleted.** They existed only to compute `log P(".") − log P(" next")` for this stage, and nothing called them once `SentenceJoin` stopped. `SentenceProbe.at`, `Slot`, `top`, `SentenceModel`, `BPETokenizer`, the ModernBERT cache directory and `check-tokenizer.sh` are all untouched — `SlotGate` and `SlotReference` still read them, and #261 is what moves that side. Say the word and I will put `read()` back.

**Not changed: `PipelineCommand.swift:196`.** `--pipeline --warm` loads ModernBERT for the vocabulary slot gate, which still needs it. The sentence join is not a pipeline stage, so nothing there wants the readings.

</details>

<details>
<summary>What the fixture and the two check scripts do now</summary>

`tests/sentence-boundary-cases.json` is rebuilt: 40 boundaries, 20 real endings and 20 cuts, each with the per-token score of every reading and the winner. It is generated by `scripts/sentence-probe-reference.py readings`, which runs the release binary's `--sentence-probe --bench` and stores what came back, so the fixture is by construction what the binary produces. Real endings carrying a hand label (`join`, `drop`, `tie`) are left out — none of the three is a real ending a join would spoil.

`scripts/check-sentence-probe.sh` compares the three scores and the winner, case by case, in one loaded process. 40/40, worst gap 0.0000.

`scripts/check-sentence-join.sh` glues each case back into one text, runs `--sentence-join`, and checks the scores and the decisions. 19 of 20 cuts repaired, 0 of 20 real periods joined, worst gap 0.0000. A joined real period fails the run on its own.

Neither is in CI: they need the 320 MB model. Both were run by hand for this PR, along with `check-tokenizer.sh` (19/19), `check-sentence-case.sh` (31/31, and it is in CI) and `check-sentence-open.sh`.

`--sentence-probe --bench <cases.json> --out <scores.json>` is new. One process per boundary pays the 1.3 s load every time, so it is the only way to get a latency number or to score a whole bench. `--vectors` loads the word vectors as well, so the memory line describes the process the app runs.

</details>

<details>
<summary>Review notes: the shared loader, and one flaky local check</summary>

The first commit is a refactor with no behaviour change. `WordVectors` held the download, the `flock`, the staging directory, the de-duplicated progress and the `FolderTokenizer` adapter. Two models need all of that now, so it is `MLXModelCache`. Same repository, same file list, same cache directory, same lock path. `--word-vector` was run before and after.

`make test` passes. One case, `every stage's variables survive the stages after it`, fails about one run in four **on this machine**. The failing dump shows `vocabulary.gate_ms = 186`, so the vocabulary gate's model ran and changed `vocabulary.count`. CI has no models, so the gate stands aside there and the case is deterministic. This PR does not touch the vocabulary gate. Flagging it rather than chasing it; CI is the check that matters.

</details>

- [x] `make test` passes.
- [x] A prompt or pattern change is scored, and the numbers are above.
- [x] Every commit is signed off — `git commit -s`. See [CONTRIBUTING.md](../CONTRIBUTING.md).

### CODE REVIEW

File: Sources/ParrotFlow/Config.swift [hunk 1]
```diff
@@ -1687,17 +1695,28 @@ struct Config: Decodable, Equatable {
...
+                    // every boundary and every period in the transcript goes.
+                    guard !kept.isEmpty else {
+                        throw ConfigError.invalidValue(
+                            key: "transcription.sentences.marks",
+                            value: "an empty list",
+                            expected: "at least one mark — with none, joining is the only"
+                                + " reading and every period would be removed"
+                        )
+                    }
+                    marks = kept
```
[the remark this answers is not in the document]
Comment (znat) [reply Sources/ParrotFlow/Config.swift:1724]: Fixed in c229886. Each entry in `marks:` must now be a single punctuation character, so `join` is refused at config load with the reason. Verified against a real config: `["join"]` and `["--"]` are refused by `--check-config`, `[".", ","]` and `[".", "?", "!"]` are accepted.

I took the reject option rather than encoding the key. A mark is written into the sentence as punctuation, so a word there was never valid — refusing it says that, where an encoded key would silently accept `marks: ["join"]` and read it as a word.
